### PR TITLE
Split daml3-script into internal and stable package

### PR DESCRIPTION
**Pull request by samuel-williams-da**
The goal here is to move "exported" types/exceptions from daml script into a separate stable package, then force the rest of the types to be non serializable. We achieve the latter with a new `--force-utility-package` flag, which ensures the generated package will not be considered for upgrades.
Note that all exceptions must be Stable, but not all newtypes/datatypes. However, since the Stable package must still compile, any types that exceptions depend on must also be stable.

In total, the stable types are:
- `data PartyManagement.PartyDetails`
- `newtype PartyManagement.PartyIdHint`
- `newtype PartyManagement.ParticipantName`
- `newtype UserManagement.UserId`
- `data UserManagement.User`
- `data UserManagement.UserRight`
- `exception UserManagement.InvalidUserId`
- `exception UserManagement.UserAlreadyExists`
- `exception UserManagement.UserNotFound`
- `newtype Testing.CommandName` <- required by Testing.FailedCmd below
- `newtype Testing.ErrorClassName` <- required by Testing.FailedCmd below
- `newtype Testing.ErrorMessage` <- required by Testing.FailedCmd below
- `exception Testing.FailedCmd`

I have also moved daml3-script over to our bazel `daml-compile`. This required some tweaks to the rule, as it was putting source files in odd locations, which didn't match what daml-script had done before.

Previously, it would create a daml.yaml at root, then the source file paths would include their full bazel path (i.e. `daml-script/daml3/Daml/Script.daml`), whereas our goal would be simply `daml/Daml/Script.daml`
To achieve this, I first fixed the `daml_compile` rule to remove the label's package path from the source file, then I added a `daml_source` field, which we set to `daml`, to be consistent with how script used to be built.
This change has affected the independence check, as the source files within that dar are now correctly placed. Similarly, since dalf's include source locations with their full path, the top level dalf's hash has changed.

### CODE REVIEW

File: sdk/compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs
Comment (samuel-williams-da) [inline sdk/compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs:1]: Changes to this module follow the PR from #20354

File: sdk/rules_daml/daml.bzl [hunk 1]
```diff
@@ -138,7 +149,7 @@ def _daml_build_impl(ctx):
             cp_srcs = "\n".join([
                 make_cp_command(
                     src = src.path,
-                    dest = "$tmpdir/" + src.path,
+                    dest = "$tmpdir/" + daml_source + src.path.removeprefix(ctx.label.package + "/"),
```
Comment (samuel-williams-da) [inline sdk/rules_daml/daml.bzl:152]: To maintain the existing source paths of `daml/Daml/Script/...`, we allow an explicit daml_source flag in daml_compile. which will be put in paths and in the daml.yaml (so it makes it into the ghc inserted locations in LF)

Comment (samuel-williams-da) [reply sdk/rules_daml/daml.bzl:152]: We also remove the existing ctx.label.package prefixes, which was putting all daml files under their full bazel path in the dars.

File: sdk/compiler/damlc/tests/platform-independence.dar-hash
Comment (samuel-williams-da) [inline sdk/compiler/damlc/tests/platform-independence.dar-hash:1]: Some of the daml file paths have changed, as we no longer include the full bazel path in the generated dars from `daml_compile`

**Comment by samuel-williams-da** [comment]
For daml2-script, we need to discuss which of the 2 following options is reasonable:
- Do not split the package, do not support using it in LF1.17, emit a warning if compiling with LF1.17 and using daml2-script
- Split daml2-script, but reuse the same serializable package as daml3-script, so that a user may safely move from daml2-script to daml3-script in an upgrade.

To be discussed with @paulbrauner-da :)

File: sdk/daml-script/daml3/internal/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+load("//rules_daml:daml.bzl", "daml_compile")
+
+# These packages contain all internal logic and data types for daml-script.
+# They are forced to be utility packages (i.e. have no serializable types) for ease of development in future
+[
+    daml_compile(
+        name = "daml3-script-{}".format(lf_version),
+        srcs = glob(["**/*.daml"]),
+        # Force daml files to be under `daml` directory, to match previous behaviour
+        daml_source = "daml",
```
Comment (samuel-williams-da) [inline sdk/daml-script/daml3/internal/BUILD.bazel:15]: Rename to `generated_daml_source_directory`

**Review by dylant-da — APPROVED** [review]
Reviewed together in a call, approved pending the bazel rule field rename

**Pull request merged**